### ISSUE TITLE
Portable exe built with pkg failed to run

### ISSUE THREAD

**Issue by pixelpicosean** [RobLoach/node-raylib#74]
OS: windows 10
Node: 12.18.2

It keeps saying: `Error: A dynamic link library (DLL) initialization routine failed.` no matter which tool I use (tried pkg, nexe, electron and nw.js)

**Comment by RobLoach** [comment]
I've found the compiled raylib file had to be in the same directory. There's some similar issues in the pkg queue too...

**Comment by pixelpicosean** [comment]
Still not work even I put `node-raylib.node`, `node-raylib.exp`, `node-raylib.lib` all in the same directory of exe.

**Comment by 77Z** [comment]
@pixelpicosean I have the same issue, I'm going to try a couple more things however.

**Comment by 77Z** [comment]
Yeah.. the only thing I could find is putting the .node file next the the executable, but that's it.

**Comment by mordof** [comment]
OS: windows 11
Node: 22.17.0

Not sure if it's useful putting this as a comment, or making a new ticket...

I'm running into the same error, but I'm trying to use Node's SEA for building the exe.  I've got rollup working for all other modules and files into one, but the .node file for raylib needs to be separate for the require to load.  

I've tied putting the .node file beside the .exe, I've tried putting all of the files from the raylib node module beside the .exe, i've tried mirroring both the node_modules folder path with raylib beside the .exe, i've tried just the contents of the node_modules/raylib folder beside the .exe, none of which changed the behavior. I encountered `Error: A dynamic link library (DLL) initialization routine failed.` for every situation.

Is there any additional information anyone has found on how this might be resolved?

**Comment by konsumer** [comment]
I looked into it a bit more, and it seems like [yao-pkg](https://github.com/yao-pkg/pkg) fork is more maintained.

I made [this tester](https://github.com/konsumer/node-raylib-pkg-example) that follows their directions, and I got it building fine on mac. Edit `targets` to add more targets, and it should build those, too. Keep in mind, there are issues building arm64 on x86-64 (but not vice-versa) so mind that, but otherwise it should work.

**Comment by mordof** [comment]
Here's the github repo I put together for submitting the node issue:
https://github.com/mordof/node_sea_crash_example

And here's the nodejs issue link:
https://github.com/nodejs/node/issues/59574

I'm a long way off from needing this to actually work, so I'm intentionally being stubborn about trying to use later/latest node versions and the SEA setup it has 😆

Since `require()` isn't allowed in a SEA bundle, I'm using `process.dlopen()` instead.  this is what gets called under the hood anyway when it detects a .node module is being called with `require()`.  Any process that uses a tmp file to write the contents, and loads it in, I've found has used dlopen() on the tmp file after as well.  In the `src/original.js` and `src/fresh_compile.js` I simply used a hoisted function that has the dlopen call in it which brings in node-raylibs wrapper code so that everything functions as expected.

I haven't looked into other platforms at all yet (though I imagine I'll want to later on). got fully stuck on this since I'm on windows 😓 

I'm gonna bookmark this thread so I'll have all the extra information you've provided here for referencing, very valuable info thank you very much.

**Comment by konsumer** [comment]
> Since require() isn't allowed in a SEA bundle, I'm using process.dlopen() instead.

I think a few things are bing conflated here.

pkg manages all this with assets, so if you put the .node file in your assets, it will work. If you fork my repo, you should get a working setup, even on windows. Your original question was about pkg, and the repo I made should work fine for that. You have to use require because the old versions of node they use don't support ESM, and pkg has some stuff hardcoded for `require`.

SEA does much less automatically, but is probably the way forward, once all the work has been done to support things. Calling `dlopen` directly might seem like it should work, but there are a few more steps, and as far as I understand SEA does not really support native modules at all (yet.) You can test this by making a much simpler native module, and it should fail, as far as I know (I last messed with SEA about a year ago, so it may have progressed.)

Additionally, [bunjs](https://bun.com/) works with build:

```sh
bun build --compile --outfile=dist/buntest ./test.js
```

I would suggest using pkg with node18/commonjs or bun, since it has pretty good native-node-module support, at this point.

**Comment by mordof** [comment]
Oh, yeah the conflated pieces are due to me hijacking someone else's thread 😓. Wasn't sure if I should've started a new issue (looks like I should have).

The original question about pkg was from the OP (different person).

I did try both bun and deno, and neither worked with the node-raylib module 🤔  Especially when looking at deno, the node api isn't implemented enough to make use of, and bun is missing some of the more detailed functionality for node-raylib to work (I want to fix the bindings that utilize arrays, for example).

If I end up getting far enough in my project that I need to have bundled exes working, and SEA hasn't caught up yet, then I will look at downgrading node versions to use pkg at that time.

**Comment by konsumer** [comment]
> The original question about pkg was from the OP (different person).

Ah, ok. well, the short answer to your not-same-but-adjancent issue is "SEA is not there yet, but maybe it will be eventually, but you can do it with pkg or bun, now"

> I did try both bun and deno, and neither worked with the node-raylib module 🤔

try it again with latest bun. I tested from [my test repo](https://github.com/konsumer/node-raylib-pkg-example) just now, and it works fine. I can't really speak on deno, but bun has nutso feature-velocity, and a serious drive to get 100% compatibility with node, but be better. I have been watching them add enough support to run node-raylib demos, and it's basically working ok, now, without any in-depth testing.

> downgrading node versions to use pkg at that time.

you don't need to downgrade node, just don't use post-node-18 syntax in your code, with pkg. In my exmample, I am using my normal node v23.11.0, but it downloads the node18 runtime stub when it compiles the executable.

**Comment by mordof** [comment]
> Ah, ok. well, the short answer to your not-same-but-adjancent issue is "SEA is not there yet, but maybe it will be eventually, but you can do it with pkg or bun, now"

Helpful to know :) 

> try it again with latest bun. I tested from [my test repo](https://github.com/konsumer/node-raylib-pkg-example) just now, and it works fine. I can't really speak on deno, but bun has nutso feature-velocity, and a serious drive to get 100% compatibility with node, but be better. I have been watching them add enough support to run node-raylib demos, and it's basically working ok, now, without any in-depth testing.

oooh ok, I'll mark that down for later. 

> you don't need to downgrade node, just don't use post-node-18 syntax in your code

Oh nice! Also just occurred to me I can set a bundler target to node 18 and have it convert my more recent code (I imagine there'll be stubs for most things to bump features down that I ended up using). Either way - that's helpful that I don't need node 18 installed specifically.


You've been a fantastic help sorting out the current state of things, and providing options. Very much appreciated 👍

**Comment by mordof** [comment]
@twuky  What!! that's an amazing bit of information, thank you!

> Coming to this way way late, and it's been a few years (wow!) but I am remembering the issues with this being solved on windows by renaming the compiled .exe to node.exe rather than your app name ??? We also used to use a bundler called Caxa but this seems discontinued.

this actually worked for both the original node-raylib released .node module, as well as my fresh compiled from from master.  renaming the finished .exe to `node.exe` and everything works.  (even the fresh compiled one that changed from a failed to init dll to a silent full crash - that worked in being named `node.exe`).

> I remember also wanting to try to see if we could fix this by messing around with [process.title](https://nodejs.org/api/process.html#processtitle) right before the app tries to require the native module?

Did you end up getting any info/hints if this (or any other possible things to investigate) might be helpful? I'd be willing to give things a try

**Comment by twuky** [comment]
worst case scenario you have to build a custom app that turns around and runs the node.exe hahaha. may be an easier way to handle custom file icons + signing the executable with MS ;)

**Comment by twuky** [comment]
> cmake-js will add the hook if the CMakeLists.txt contains the library `${CMAKE_JS_SRC}`

Looks like we're not including this in our current cmakelists, not sure what it needs (linking this in the build too?) but could be worth adding

> Ideally I think it'd be fantastic if node-raylib could end up having either a config file, or package.json entry, or something that would allow a user to clone the repo, change the exe config name they want, and have that be utilized in the code gen/compile process (whichever is needed) to allow the user to have a compiled module for their target exe.

If what I'm understanding about the way this fix works (adding some delay so that windows only trys to load the dll after the process name is registered correctly) it probably shouldnt matter what your filename ends up being
